File: src/de/saschahlusiak/freebloks/view/SimpleModel.java
package de.saschahlusiak.freebloks.view;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.FloatBuffer;
import java.nio.ShortBuffer;

import javax.microedition.khronos.opengles.GL10;

public class SimpleModel {
	int num_triangles;
	
    private ShortBuffer _indexBuffer;
    private FloatBuffer _vertexBuffer;
    private FloatBuffer _normalBuffer;
    private FloatBuffer _textureBuffer;
        
    public SimpleModel(int num_vertices, int num_triangles) {
    	this.num_triangles = num_triangles;
    	
	    // float has 4 bytes
	    ByteBuffer vbb = ByteBuffer.allocateDirect(num_vertices * 3 * 4);
	    vbb.order(ByteOrder.nativeOrder());
	    _vertexBuffer = vbb.asFloatBuffer();
	    
	    vbb = ByteBuffer.allocateDirect(num_vertices * 3 * 4);
	    vbb.order(ByteOrder.nativeOrder());
	    _normalBuffer = vbb.asFloatBuffer();
	 
	    // short has 2 bytes
	    ByteBuffer ibb = ByteBuffer.allocateDirect(num_triangles * 3 * 2);
	    ibb.order(ByteOrder.nativeOrder());
	    _indexBuffer = ibb.asShortBuffer();
	    
	    ByteBuffer tbb = ByteBuffer.allocateDirect(num_vertices * 2 * 4);
	    tbb.order(ByteOrder.nativeOrder());
	    _textureBuffer = tbb.asFloatBuffer();
    }
    
    public void commit() {
	    _vertexBuffer.position(0);
	    _normalBuffer.position(0);
	    _indexBuffer.position(0);
	    _textureBuffer.position(0);   
   	
    }
    
	public void addVertex(float x, float y, float z, float nx, float ny, float nz, float tu, float tv) {
		_vertexBuffer.put(x);
		_vertexBuffer.put(y);
		_vertexBuffer.put(z);
		_normalBuffer.put(nx);
		_normalBuffer.put(ny);
		_normalBuffer.put(nz);
		_textureBuffer.put(tu);
		_textureBuffer.put(tv);
	}

	public void addIndex(int v1, int v2, int v3) {
		_indexBuffer.put((short)(v1));
		_indexBuffer.put((short)(v2));
		_indexBuffer.put((short)(v3));
	}
	
	public FloatBuffer getVertexBuffer() {
		return _vertexBuffer;
	}
	public FloatBuffer getNormalBuffer() {
		return _normalBuffer;
	}
	public FloatBuffer getTextureBuffer() {
		return _textureBuffer;
	}
	public ShortBuffer getIndexBuffer() {
		return _indexBuffer;
	}

	public void drawElements(GL10 gl) {
		gl.glDrawElements(GL10.GL_TRIANGLES, num_triangles * 3, GL10.GL_UNSIGNED_SHORT, getIndexBuffer());
	}
}

File: src/main/java/de/saschahlusiak/freebloks/view/SimpleModel.kt
package de.saschahlusiak.freebloks.view

import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.ShortBuffer

import javax.microedition.khronos.opengles.GL10
import javax.microedition.khronos.opengles.GL11

open class SimpleModel(numVertices: Int, private val numTriangles: Int, private val useVBO: Boolean = true) {
    // short has 2 bytes
    internal val indexBuffer: ShortBuffer = ByteBuffer.allocateDirect(numTriangles * 3 * 2)
        .order(ByteOrder.nativeOrder())
        .asShortBuffer()

    // float has 4 bytes
    internal val vertexBuffer = ByteBuffer.allocateDirect(numVertices * 8 * 4)
        .order(ByteOrder.nativeOrder())
        .asFloatBuffer()

    internal var vbo: IntArray? = null

    fun commit() {
        vertexBuffer.position(0)
        indexBuffer.position(0)
    }

    fun addVertex(x: Float, y: Float, z: Float, nx: Float, ny: Float, nz: Float, tu: Float, tv: Float) {
        vertexBuffer.apply {
            put(x)
            put(y)
            put(z)
            put(nx)
            put(ny)
            put(nz)
            put(tu)
            put(tv)
        }
    }

    fun addIndex(v1: Int, v2: Int, v3: Int) {
        indexBuffer.apply {
            put(v1.toShort())
            put(v2.toShort())
            put(v3.toShort())
        }
    }

    fun addIndex(offset: Int, v1: Int, v2: Int, v3: Int) {
        addIndex(v1 + offset, v2 + offset, v3 + offset)
    }

    fun bindBuffers(gl: GL11) {
        if (!useVBO) {
            gl.glBindBuffer(GL11.GL_ARRAY_BUFFER, 0)
            gl.glBindBuffer(GL11.GL_ELEMENT_ARRAY_BUFFER, 0)

            gl.glVertexPointer(3, GL10.GL_FLOAT, 4 * 8, vertexBuffer.position(0))
            gl.glNormalPointer(GL10.GL_FLOAT, 4 * 8, vertexBuffer.position(3))
            gl.glTexCoordPointer(2, GL10.GL_FLOAT, 4 * 8, vertexBuffer.position(6))
        } else {
            val vbo = vbo ?: makeVBO(gl)

            gl.glBindBuffer(GL11.GL_ARRAY_BUFFER, vbo[0])

            gl.glVertexPointer(3, GL10.GL_FLOAT, 4 * 8, 0)
            gl.glNormalPointer(GL10.GL_FLOAT, 4 * 8, 3 * 4)
            gl.glTexCoordPointer(2, GL10.GL_FLOAT, 4 * 8, 6 * 4)

            gl.glBindBuffer(GL11.GL_ELEMENT_ARRAY_BUFFER, vbo[1])
        }
    }

    fun invalidateBuffers(gl: GL11) {
        vbo = if (useVBO)
            makeVBO(gl)
        else
            null
    }

    fun drawElements(gl: GL11, mode: Int) {
        if (useVBO)
            gl.glDrawElements(mode, numTriangles * 3, GL10.GL_UNSIGNED_SHORT, 0)
        else
            gl.glDrawElements(mode, numTriangles * 3, GL10.GL_UNSIGNED_SHORT, indexBuffer)
    }

    private fun makeVBO(gl: GL11): IntArray {
        this.vbo?.let {
            gl.glDeleteBuffers(it.size, it, 0)
        }

        val vbo = IntArray(2)
        gl.glGenBuffers(2, vbo, 0)

        gl.glBindBuffer(GL11.GL_ARRAY_BUFFER, vbo[0])
        gl.glBufferData(GL11.GL_ARRAY_BUFFER,
            vertexBuffer.capacity() * 4,
            vertexBuffer.rewind(),
            GL11.GL_STATIC_DRAW)
        gl.glBindBuffer(GL11.GL_ARRAY_BUFFER, 0)

        gl.glBindBuffer(GL11.GL_ELEMENT_ARRAY_BUFFER, vbo[1])
        gl.glBufferData(GL11.GL_ELEMENT_ARRAY_BUFFER,
            indexBuffer.capacity() * 2,
            indexBuffer,
            GL11.GL_STATIC_DRAW)
        gl.glBindBuffer(GL11.GL_ELEMENT_ARRAY_BUFFER, 0)

        this.vbo = vbo

        return vbo
    }
}
